### PR TITLE
Add Plotly plots to DiffusionMap

### PR DESCRIPTION
**Pull request by lilyminium**
Is MDAnalysis interested in supporting optional Plotly graphics in DiffusionMap? [(e.g. in this notebook, some scrolling required)](https://lilyminium.github.io/UserGuide/examples/analysis/reduced_dimensions/diffusion_map.html#Passing-your-own-distance-metric-to-DiffusionMap) It's convenient for quickly looking through the mapping at different time scales. 

Changes made in this Pull Request:
 - Added 2D and 3D animated plots
 - Added frame checking/subsetting to `DiffusionMap.run()`. It's not clear that if the user `run()`s `DistanceMatrix` themselves, that they cannot alter the frames of the analysis in `DiffusionMap.run()`
 - Added NumPy arrays to `DiffusionMap`. `DiffusionMap` doesn't actually operate on a universe or trajectory, just the distance matrix thereof. If I have my own distance matrix, I would like to use `DiffusionMap` on it instead of writing my own analysis.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 1]
```diff
@@ -364,10 +365,191 @@ def transform(self, n_eigenvectors, time):
...
                 (self.eigenvalues[1:n_eigenvectors+1]**time))
+
+    def plot_animated_transform(self, n_time=25, colorscale='Viridis', marker_size=3,
+                                x=1, y=2):
+        """ Plots an interactive 2D plot of chosen eigenvectors at different timescales.
+
+        Parameters
+        ----------
+        n_time: int, optional
+            number of time values for ``transform()``. Default: 25
```
Comment (RMeli) [inline package/MDAnalysis/analysis/diffusionmap.py]: I think the default parameters are automatically included in the documentation from the function definition. Should we avoid specifying them in the docstring as well, in order to avoid possible divergences between documentation and code?

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 2]
```diff
@@ -364,10 +365,191 @@ def transform(self, n_eigenvectors, time):
...
+
+
+        .. versionadded:: 1.0.0
+        """
+        try:
+            import plotly.graph_objects as go
+        except ImportError:
+            raise ImportError('Plotly is required for this animated graph')
+
+        if n_time is None:
```
Comment (RMeli) [inline package/MDAnalysis/analysis/diffusionmap.py]: `n_frames` instead of `n_time` cold be a clearer name?

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 3]
```diff
@@ -364,10 +365,191 @@ def transform(self, n_eigenvectors, time):
...
+
+    def plot_animated_transform3D(self, n_time=25, colorscale='Viridis', marker_size=3,
+                                  x=1, y=2, z=3):
+        """
+        Plots an interactive 3D plot of chosen eigenvectors at different timescales.
+
+        Parameters
+        ----------
+        n_time: int, optional
+            number of time values for ``transform()``. Default: 25
```
Comment (RMeli) [inline package/MDAnalysis/analysis/diffusionmap.py]: See comment above.

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 4]
```diff
@@ -364,10 +365,191 @@ def transform(self, n_eigenvectors, time):
...
+        """
+        try:
+            import plotly.graph_objects as go
+        except ImportError:
+            raise ImportError('Plotly is required for this animated graph')
+
+        eig = np.array([x, y, z]) - 1
+        n_eig = max(eig) + 1
+
+        if n_time is None:
```
Comment (RMeli) [inline package/MDAnalysis/analysis/diffusionmap.py]: See comment above.

File: package/MDAnalysis/analysis/diffusionmap.py (hunk 1 above)
Comment (lilyminium) [reply package/MDAnalysis/analysis/diffusionmap.py]: Yes, I reckon so. I think I just do it because [everyone](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) [else](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html) [seems](https://numpy.org/doc/1.18/reference/generated/numpy.trace.html#numpy.trace) [to](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.apply.html#pandas.DataFrame.apply), but that's not a very good reason -- maybe it's from when `**kwargs` was everywhere.

File: package/MDAnalysis/analysis/diffusionmap.py (hunk 2 above)
Comment (lilyminium) [reply package/MDAnalysis/analysis/diffusionmap.py]: I've changed it to `n_frames` for now because it *does* pass integers representing frames to ``transform()``, but I'm not sure that doesn't make it more confusing given that the keyword for ``transform`` is ``time``. What do you think of docstring/s?

**Comment by lilyminium** [comment]
Added bits to deal with raw numpy arrays and start/stop/step mismatches in `run()`; updated the first comment to explain why.

**Comment by hmacdope** [comment]
@lilyminium do we still want this ? Otherwise close.?

**Comment by lilyminium** [comment]
Yea let’s close, at this point I think the last thing we want is more dependencies!On Oct 7, 2023, at 16:44, Hugo MacDermott-Opeskin ***@***.***> wrote:﻿
@lilyminium do we still want this ? Otherwise close.?
